Add WTF and bmalloc modulemaps

adetaylor opened the pull request:
#### f5eaad8b6448439dc390ee75f04a6395f71dabb2
<pre>
Add WTF and bmalloc modulemaps
<a href="https://bugs.webkit.org/show_bug.cgi?id=292791">https://bugs.webkit.org/show_bug.cgi?id=292791</a>
<a href="https://rdar.apple.com/151023888">rdar://151023888</a>

Reviewed by Elliott Williams.

This commit adds modulemap files for WTF and bmalloc. These are necessary
because Swift/C++ interop interprets C++ headers using the &quot;-fmodules&quot; option,
interpreting the C++ codebase as clang modules.

The approach taken for the two modulemaps is slightly different. bmalloc
defines a single monolithic module, the simplest approach. For WTF we choose
a more complex approach, as previously followed by the WebKit_Internal
module, where the module is divided into many small submodules. This
approach is chosen because Swift/C++ interoperability otherwise tries
to interpret the whole module, and that results in errors such as:

  /usr/include/c++/v1/__memory/construct_at.h:47:58: note: candidate template ignored: substitution failure [with _Tp = WTF::TextBreakIterator, _Args = &lt;const WTF::TextBreakIterator &amp;&gt;]: call to deleted constructor of &apos;WTF::TextBreakIterator&apos;
_LIBCPP_HIDE_FROM_ABI _LIBCPP_CONSTEXPR_SINCE_CXX20 _Tp* __construct_at(_Tp* __location, _Args&amp;&amp;... __args) {

We&apos;ll likely need to solve those in future, but meanwhile splitting the
module into lots of submodules limits the blast radius of such issues
and allows us to make progress with Swift.

Notes on some specific parts of the WTF modulemap:
* HashMap and RobinHoodHashTable.h are interdependent so need to be
  colocated in a module.
* The Platform headers are used in contexts without C++11 support, unlike
  other parts of WTF.

Although it just adds files, this commit is not a no-op because some of
our existing Swift build targets will spot these modulemaps and interpret
headers in a modularized style. Nevertheless, there are no
intentional behavior changes here.

* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/module.modulemap: Added.
* Source/bmalloc/bmalloc.xcodeproj/project.pbxproj:
* Source/bmalloc/module.modulemap: Added.

Canonical link: <a href="https://commits.webkit.org/294864@main">https://commits.webkit.org/294864@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c24eb859cf2817f90ae95c9110112ac2a1db7e42

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/103321 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/22997 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/13318 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/108494 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/53964 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/23337 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/31548 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/78521 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/35456 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/106327 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/18090 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/93209 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/58854 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/17928 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/11250 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/53320 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/95996 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/87729 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/11312 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/110869 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/101932 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/30459 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/22491 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/87524 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/30824 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/89406 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/87160 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22186 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/32015 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/9733 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/24793 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/30388 "Built successfully") | [❌ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/35707 "Failed to build and analyze WebKit") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/125565 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/30194 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/34814 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/33521 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/31756 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

emw-apple commented on Source/WTF/wtf/module.modulemap:
```diff
@@ -0,0 +1,546 @@
+module wtf {
+    config_macros UCHAR_TYPE
+    explicit module AbstractRefCounted { requires cplusplus11 header "AbstractRefCounted.h" }
```
How was `requires cplusplus11` chosen? We're on c++23, so I doubt it's accurate.

I'd be in favor of removing this directive unless we need it; WTF doesn't have non-WebKit clients, so there's little use in making the module resilient to importing in others projects' builds.

adetaylor replied:
`libWebKitSwift` does not specify any particular C++ version (which is fair enough). It attempts to build `wtf` with a modulemap, without specifying any C++ version, and thus many WTF headers fail to find C++11 headers such as `<type_traits>`. Error:

```
error: Clang dependency scanner failure: While building module 'AuthenticationServicesCore_Private' imported from AuthenticationServicesCore_Private-5bc0c8cb.input:1:
While building module 'AuthenticationServicesCore' imported from /Applications/Xcode2.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX15.5.Internal.sdk/System/Library/PrivateFrameworks/AuthenticationServicesCore.framework/PrivateHeaders/_ASCABLEPowerLogger.h:3:
While building module 'SafariCore' imported from /Applications/Xcode2.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX15.5.Internal.sdk/System/Library/PrivateFrameworks/AuthenticationServicesCore.framework/Headers/ASCPasswordLoginChoice.h:6:
While building module 'wtf' imported from /Applications/Xcode2.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX15.5.Internal.sdk/System/Library/PrivateFrameworks/SafariCore.framework/Headers/WBSCoreFeatureDefinitions.h:13:
In file included from <module-includes>:2:
In file included from /Users/adrian_taylor/WebKit/OpenSource/WebKitBuild/Debug/usr/local/include/wtf/AbstractRefCountedAndCanMakeWeakPtr.h:29:
In file included from /Users/adrian_taylor/WebKit/OpenSource/WebKitBuild/Debug/usr/local/include/wtf/CanMakeWeakPtr.h:29:
In file included from /Users/adrian_taylor/WebKit/OpenSource/WebKitBuild/Debug/usr/local/include/wtf/CompactRefPtrTuple.h:28:
/Users/adrian_taylor/WebKit/OpenSource/WebKitBuild/Debug/usr/local/include/wtf/CompactPointerTuple.h:29:10: fatal error: 'type_traits' file not found
While building module 'AuthenticationServicesCore_Private' imported from AuthenticationServicesCore_Private-5bc0c8cb.input:1:
While building module 'AuthenticationServicesCore' imported from /Applications/Xcode2.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX15.5.Internal.sdk/System/Library/PrivateFrameworks/AuthenticationServicesCore.framework/PrivateHeaders/_ASCABLEPowerLogger.h:3:
While building module 'SafariCore' imported from /Applications/Xcode2.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX15.5.Internal.sdk/System/Library/PrivateFrameworks/AuthenticationServicesCore.framework/Headers/ASCPasswordLoginChoice.h:6:
In file included from <module-includes>:1:
In file included from /Applications/Xcode2.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX15.5.Internal.sdk/System/Library/PrivateFrameworks/SafariCore.framework/Headers/SafariCore.h:4:
In file included from /Applications/Xcode2.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX15.5.Internal.sdk/System/Library/PrivateFrameworks/SafariCore.framework/Headers/WBSAccessibilityUtilities.h:5:
/Applications/Xcode2.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX15.5.Internal.sdk/System/Library/PrivateFrameworks/SafariCore.framework/Headers/WBSCoreFeatureDefinitions.h:13:9: fatal error: could not build module 'wtf'       
```

I could almost certainly specify `cplusplus23` instead. `cplusplus11` works to suppress the above error message, because then nothing attempts to understand these modules with a pre-C++11 version.

adetaylor replied:
I tried `cplusplus23` and it doesn't build nicely, possibly because we are currently using `c++2b`?

emw-apple replied:
> `libWebKitSwift` does not specify any particular C++ version (which is fair enough). It attempts to build `wtf` with a modulemap, without specifying any C++ version, and thus many WTF headers fail to find C++11 headers such as `<type_traits>`.

Hm, that's a little surprising to me since we set `CLANG_CXX_LANGUAGE_STANDARD = c++2b;` in WebKit's `Base.xcconfig`. Maybe because it's not building with C++ interop? (i.e. it needs `SWIFT_OBJC_INTEROP_MODE = objcxx`).

I'd prefer to not have another place where we declare a supported c++ standard, if at all possible.

adetaylor replied:
OK. I think that's rdar://149474443 "Swift/C++ interop does not respect CLANG_CXX_LANGUAGE_STANDARD = "c++2b";". I have worked around that bug in other places using `OTHER_SWIFT_FLAGS = $(inherited) -Xcc -std=c++2b` in `WebKit.xcconfig`.

I've found that I can't apply `OTHER_SWIFT_FLAGS = $(inherited) -Xcc -std=c++2b` in (for instance) `Base.xcconfig` because it comes to apply even to projects which are only trying to import obj-c, not c++, and Swift's internal clang then understandably complains.

So until that bug is fixed, I think the provision of `requires cplusplus11` in each module is the best option. If you have other suggestions I'm all ears!

emw-apple replied:
How about this in Base.xcconfig and elsewhere?
```
OTHER_SWIFT_FLAGS = $(inherited) $(OTHER_SWIFT_FLAGS_$(SWIFT_OBJC_INTEROP_MODE));
OTHER_SWIFT_FLAGS_objcxx = -Xcc -std=$(CLANG_CXX_LANGUAGE_STANDARD);
```

adetaylor replied:
Good idea. I'll give it a try. It might not work if there are any projects which import from both obj-c and c++ headers, and enable objcxx interop mode, but hopefully there aren't any such projects.

adetaylor replied:
It doesn't work, because the actual errors are from `libWebKitSwift` (no C++ interop enabled), depending on private frameworks, which in turn depend upon WTF.

```
error: Clang dependency scanner failure: While building module 'AuthenticationServicesCore_Private' imported from AuthenticationServicesCore_Private-734076b4.input:1:
While building module 'AuthenticationServicesCore' imported from /Applications/Xcode2.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX15.5.Internal.sdk/System/Library/PrivateFrameworks/AuthenticationServicesCore.framework/PrivateHeaders/_ASCABLEPowerLogger.h:3:
While building module 'SafariCore' imported from /Applications/Xcode2.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX15.5.Internal.sdk/System/Library/PrivateFrameworks/AuthenticationServicesCore.framework/Headers/ASCPasswordLoginChoice.h:6:
While building module 'wtf' imported from /Applications/Xcode2.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX15.5.Internal.sdk/System/Library/PrivateFrameworks/SafariCore.framework/Headers/WBSCoreFeatureDefinitions.h:13:
In file included from <module-includes>:2:
In file included from /Users/adrian_taylor/WebKit/OpenSource/WebKitBuild/Debug/usr/local/include/wtf/AbstractRefCountedAndCanMakeWeakPtr.h:29:
In file included from /Users/adrian_taylor/WebKit/OpenSource/WebKitBuild/Debug/usr/local/include/wtf/CanMakeWeakPtr.h:29:
In file included from /Users/adrian_taylor/WebKit/OpenSource/WebKitBuild/Debug/usr/local/include/wtf/CompactRefPtrTuple.h:28:
/Users/adrian_taylor/WebKit/OpenSource/WebKitBuild/Debug/usr/local/include/wtf/CompactPointerTuple.h:29:10: fatal error: 'type_traits' file not found
While building module 'AuthenticationServicesCore_Private' imported from AuthenticationServicesCore_Private-734076b4.input:1:
While building module 'AuthenticationServicesCore' imported from /Applications/Xcode2.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX15.5.Internal.sdk/System/Library/PrivateFrameworks/AuthenticationServicesCore.framework/PrivateHeaders/_ASCABLEPowerLogger.h:3:
While building module 'SafariCore' imported from /Applications/Xcode2.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX15.5.Internal.sdk/System/Library/PrivateFrameworks/AuthenticationServicesCore.framework/Headers/ASCPasswordLoginChoice.h:6:
In file included from <module-includes>:1:
In file included from /Applications/Xcode2.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX15.5.Internal.sdk/System/Library/PrivateFrameworks/SafariCore.framework/Headers/SafariCore.h:4:
In file included from /Applications/Xcode2.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX15.5.Internal.sdk/System/Library/PrivateFrameworks/SafariCore.framework/Headers/WBSAccessibilityUtilities.h:5:
/Applications/Xcode2.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX15.5.Internal.sdk/System/Library/PrivateFrameworks/SafariCore.framework/Headers/WBSCoreFeatureDefinitions.h:13:9: fatal error: could not build module 'wtf'
While building module 'AuthenticationServicesCore_Private' imported from AuthenticationServicesCore_Private-734076b4.input:1:
While building module 'AuthenticationServicesCore' imported from /Applications/Xcode2.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX15.5.Internal.sdk/System/Library/PrivateFrameworks/AuthenticationServicesCore.framework/PrivateHeaders/_ASCABLEPowerLogger.h:3:
In file included from <module-includes>:1:
In file included from /Applications/Xcode2.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX15.5.Internal.sdk/System/Library/PrivateFrameworks/AuthenticationServicesCore.framework/Headers/AuthenticationServicesCore.h:23:
/Applications/Xcode2.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX15.5.Internal.sdk/System/Library/PrivateFrameworks/AuthenticationServicesCore.framework/Headers/ASCPasswordLoginChoice.h:6:9: fatal error: could not build module 'SafariCore'
While building module 'AuthenticationServicesCore_Private' imported from AuthenticationServicesCore_Private-734076b4.input:1:
In file included from <module-includes>:1:
In file included from /Applications/Xcode2.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX15.5.Internal.sdk/System/Library/PrivateFrameworks/AuthenticationServicesCore.framework/PrivateHeaders/AuthenticationServicesCorePrivate.h:3:
/Applications/Xcode2.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX15.5.Internal.sdk/System/Library/PrivateFrameworks/AuthenticationServicesCore.framework/PrivateHeaders/_ASCABLEPowerLogger.h:3:9: fatal error: could not build module 'AuthenticationServicesCore'
AuthenticationServicesCore_Private-734076b4.input:1:1: fatal error: could not build module 'AuthenticationServicesCore_Private'
```

So I think specifying `cplusplusSOMETHING` in the `wtf` module map is actually the architecturally _correct_ way to avoid those modules being depended upon in these contexts. It's a shame it has to be `cplusplus11`, but arguably that's correct too, since the modules _can_ resolve transitive module dependencies (e.g. `type_traits`) in a C++11 context, but not in earlier versions.

emw-apple replied:
> So I think specifying `cplusplusSOMETHING` in the `wtf` module map is actually the architecturally _correct_ way to avoid those modules being depended upon in these contexts.

This reason makes a lot of sense! The build system compiles modules ahead of time, and since we have some non-interop targets, it will compile them at least once `-x objective-c` mode, where we need to instruct it to not enter the C++ code.

adetaylor replied:
There's a bit of follow-up discussion [here](https://github.com/WebKit/WebKit/pull/51245#discussion_r2393312741) about the C++ version (and a link to a PR where we're going to try to bump it to c++20)

emw-apple reviewed (commented):
Nice!

I don't think the ergonomics of many WTF submodules are so bad. In clang, it matches existing semantics: If we adopt `-fmodules` in the future, `#include <wtf/Vector.h>` still gets the AST of Vector. This avoids potential avoid memory use/build time regressions that might be caused by changing _any_ import of WTF to import _all_ of WTF.

In Swift, I think it's valuable to require clients to be specific about which WTF headers they are using. There are plenty of WTF-isms like `RefPtr` that don't make sense in swift code, so the explicit import of these submodules is nice IMO.

We will need to write a build-time check that verifies that all of WTF and bmalloc's private headers are modularized. Doesn't need to happen now, but please file a bug for follow-up.

adetaylor committed:
Add WTF and bmalloc modulemaps
https://bugs.webkit.org/show_bug.cgi?id=292791
rdar://151023888

Reviewed by Elliott Williams.

This commit adds modulemap files for WTF and bmalloc. These are necessary
because Swift/C++ interop interprets C++ headers using the "-fmodules" option,
interpreting the C++ codebase as clang modules.

The approach taken for the two modulemaps is slightly different. bmalloc
defines a single monolithic module, the simplest approach. For WTF we choose
a more complex approach, as previously followed by the WebKit_Internal
module, where the module is divided into many small submodules. This
approach is chosen because Swift/C++ interoperability otherwise tries
to interpret the whole module, and that results in errors such as:

  /usr/include/c++/v1/__memory/construct_at.h:47:58: note: candidate template ignored: substitution failure [with _Tp = WTF::TextBreakIterator, _Args = <const WTF::TextBreakIterator &>]: call to deleted constructor of 'WTF::TextBreakIterator'
_LIBCPP_HIDE_FROM_ABI _LIBCPP_CONSTEXPR_SINCE_CXX20 _Tp* __construct_at(_Tp* __location, _Args&&... __args) {

We'll likely need to solve those in future, but meanwhile splitting the
module into lots of submodules limits the blast radius of such issues
and allows us to make progress with Swift.

Notes on some specific parts of the WTF modulemap:
* HashMap and RobinHoodHashTable.h are interdependent so need to be
  colocated in a module.
* The Platform headers are used in contexts without C++11 support, unlike
  other parts of WTF.

Although it just adds files, this commit is not a no-op because some of
our existing Swift build targets will spot these modulemaps and interpret
headers in a modularized style. Nevertheless, there are no
intentional behavior changes here.

* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/module.modulemap: Added.
* Source/bmalloc/bmalloc.xcodeproj/project.pbxproj:
* Source/bmalloc/module.modulemap: Added.

Canonical link: https://commits.webkit.org/294864@main
